File: frontend/src/services/api/knowledge.ts
/**
 * Knowledge map and knowledge-gap client.
 *
 * Reading never generates: `fetchKnowledgeMap` is a GET and comes back empty for a
 * course whose map has not been built. Generation is an explicit POST because it
 * costs model calls.
 */

import { apiRequest } from '@/services/api/client';
import type { QuizSource } from '@/services/api/learning';
import type { ISODateString } from '@/types/domain';

/* -------------------------------------------------------------------------- */
/*  Wire types                                                                 */
/* -------------------------------------------------------------------------- */

interface SourceDto {
  document_id: string;
  document_name: string;
  page_number: number | null;
  chunk_id: string;
}

interface NodeDto {
  topic_id: string;
  name: string;
  description: string;
  mastery: number;
  effective_mastery: number;
  band: string;
  band_label: string;
  questions_attempted: number;
}

interface EdgeDto {
  id: string;
  source_topic_id: string;
  target_topic_id: string;
  relationship_type: RelationshipType;
  supporting_chunk_count: number;
  sources: SourceDto[];
}

interface MapDto {
  course_id: string;
  nodes: NodeDto[];
  edges: EdgeDto[];
  generated_at: ISODateString | null;
}

interface GenerateMapDto extends MapDto {
  candidates_rejected: number;
  model_calls: number;
}

interface GapDto {
  topic_id: string;
  name: string;
  kind: GapKind;
  kind_label: string;
  severity: number;
  effective_mastery: number;
  blocked_topics: string[];
  attempted_dependents: string[];
  reason: string;
}

interface GapsDto {
  course_id: string;
  gaps: GapDto[];
  has_map: boolean;
}

/* -------------------------------------------------------------------------- */
/*  Domain types                                                               */
/* -------------------------------------------------------------------------- */

/** `prerequisite` is directed (source must come first); `related` is not. */
export type RelationshipType = 'prerequisite' | 'related';

export type GapKind = 'unmet_prerequisite' | 'blocking' | 'isolated';

export interface KnowledgeNode {
  topicId: string;
  name: string;
  description: string;
  mastery: number;
  /** Mastery after the retention heuristic — what the map colours by. */
  effectiveMastery: number;
  band: string;
  bandLabel: string;
  questionsAttempted: number;
}

export interface KnowledgeEdge {
  id: string;
  sourceTopicId: string;
  targetTopicId: string;
  relationshipType: RelationshipType;
  /**
   * How many excerpts support this link. A count of evidence we hold, NOT a
   * confidence the model reported — the UI says "supported by N excerpts" so the
   * number means exactly what it is.
   */
  supportingChunkCount: number;
  sources: QuizSource[];
}

export interface KnowledgeMap {
  courseId: string;
  nodes: KnowledgeNode[];
  edges: KnowledgeEdge[];
  generatedAt: ISODateString | null;
}

export interface KnowledgeMapGeneration extends KnowledgeMap {
  candidatesRejected: number;
  modelCalls: number;
}

export interface KnowledgeGap {
  topicId: string;
  name: string;
  kind: GapKind;
  kindLabel: string;
  /** 0–1. Deterministic; see backend/app/services/learning/knowledge.py. */
  severity: number;
  effectiveMastery: number;
  blockedTopics: string[];
  attemptedDependents: string[];
  reason: string;
}

export interface KnowledgeGaps {
  courseId: string;
  gaps: KnowledgeGap[];
  /** False when no map has been generated — gaps still work, with less context. */
  hasMap: boolean;
}

/* -------------------------------------------------------------------------- */
/*  Mapping                                                                    */
/* -------------------------------------------------------------------------- */

function toSource(dto: SourceDto): QuizSource {
  return {
    documentId: dto.document_id,
    documentName: dto.document_name,
    pageNumber: dto.page_number,
    chunkId: dto.chunk_id,
  };
}

function toNode(dto: NodeDto): KnowledgeNode {
  return {
    topicId: dto.topic_id,
    name: dto.name,
    description: dto.description,
    mastery: dto.mastery,
    effectiveMastery: dto.effective_mastery,
    band: dto.band,
    bandLabel: dto.band_label,
    questionsAttempted: dto.questions_attempted,
  };
}

function toEdge(dto: EdgeDto): KnowledgeEdge {
  return {
    id: dto.id,
    sourceTopicId: dto.source_topic_id,
    targetTopicId: dto.target_topic_id,
    relationshipType: dto.relationship_type,
    supportingChunkCount: dto.supporting_chunk_count,
    sources: dto.sources.map(toSource),
  };
}

function toMap(dto: MapDto): KnowledgeMap {
  return {
    courseId: dto.course_id,
    nodes: dto.nodes.map(toNode),
    edges: dto.edges.map(toEdge),
    generatedAt: dto.generated_at,
  };
}

function toGap(dto: GapDto): KnowledgeGap {
  return {
    topicId: dto.topic_id,
    name: dto.name,
    kind: dto.kind,
    kindLabel: dto.kind_label,
    severity: dto.severity,
    effectiveMastery: dto.effective_mastery,
    blockedTopics: dto.blocked_topics,
    attemptedDependents: dto.attempted_dependents,
    reason: dto.reason,
  };
}

/* -------------------------------------------------------------------------- */
/*  Requests                                                                   */
/* -------------------------------------------------------------------------- */

export async function fetchKnowledgeMap(courseId: string): Promise<KnowledgeMap> {
  return toMap(await apiRequest<MapDto>(`/v1/courses/${courseId}/knowledge-map`));
}

export async function generateKnowledgeMap(
  courseId: string,
): Promise<KnowledgeMapGeneration> {
  const dto = await apiRequest<GenerateMapDto>(
    `/v1/courses/${courseId}/knowledge-map`,
    { method: 'POST' },
  );
  return {
    ...toMap(dto),
    candidatesRejected: dto.candidates_rejected,
    modelCalls: dto.model_calls,
  };
}

export async function fetchKnowledgeGaps(courseId: string): Promise<KnowledgeGaps> {
  const dto = await apiRequest<GapsDto>(`/v1/courses/${courseId}/knowledge-gaps`);
  return {
    courseId: dto.course_id,
    gaps: dto.gaps.map(toGap),
    hasMap: dto.has_map,
  };
}

File: backend/app/api/v1/endpoints/knowledge.py
"""Knowledge map and knowledge-gap endpoints.

Generation is a POST because it costs model calls; reading is a GET and never
generates. A student who opens the tab on a course with no map sees an empty map
and a button, not a surprise bill.
"""

import uuid

from fastapi import APIRouter, status
from sqlalchemy.orm import Session

from app.api.deps import (
    CurrentUser,
    KnowledgeMapServiceDep,
    MasteryServiceDep,
    SessionDep,
)
from app.models import Document, DocumentChunk, TopicRelationship
from app.schemas import (
    KnowledgeEdgeRead,
    KnowledgeGapRead,
    KnowledgeGapsRead,
    KnowledgeMapGenerateResponse,
    KnowledgeMapRead,
    KnowledgeNodeRead,
    SourceRef,
)
from app.services.learning.grounding import page_number_for
from app.services.learning.knowledge import GAP_KIND_LABELS
from app.services.learning.mastery import BAND_LABELS, band_for
from app.services.learning.retention import effective_mastery
from app.services.rag.retrieval import RetrievedChunk

router = APIRouter(tags=["knowledge-map"])

_RESPONSES = {
    status.HTTP_401_UNAUTHORIZED: {"description": "Not authenticated."},
    status.HTTP_404_NOT_FOUND: {"description": "Not found for this user."},
    status.HTTP_400_BAD_REQUEST: {"description": "Not enough material."},
    status.HTTP_503_SERVICE_UNAVAILABLE: {"description": "AI provider not configured."},
}


def _sources_for(session: Session, edge: TopicRelationship) -> list[SourceRef]:
    """Resolve an edge's evidence rows to citations.

    Read from the chunk and document rows at render time, so a renamed file stays
    correct and a page number can never have been invented.
    """
    sources: list[SourceRef] = []
    for evidence in edge.evidence:
        chunk = session.get(DocumentChunk, evidence.chunk_id)
        document = session.get(Document, evidence.document_id)
        if chunk is None or document is None:
            # Source deleted since generation. Better no citation than a dangling
            # one; the count shown alongside is the number we can still show.
            continue
        reference = RetrievedChunk(
            chunk_id=chunk.id,
            document_id=document.id,
            document_name=document.filename,
            page_number=chunk.page_number,
            chunk_index=chunk.chunk_index,
            content=chunk.content,
            similarity=1.0,
        )
        sources.append(
            SourceRef(
                document_id=document.id,
                document_name=document.filename,
                page_number=page_number_for(reference, document.file_type),
                chunk_id=chunk.id,
            )
        )
    return sources


def _render_map(
    session: Session,
    mastery: MasteryServiceDep,
    user_id: uuid.UUID,
    course_id: uuid.UUID,
    topics,
    edges,
) -> tuple[list[KnowledgeNodeRead], list[KnowledgeEdgeRead]]:
    """Join the graph to the student's mastery for display."""
    states = {topic.id: state for topic, state in mastery.for_course(user_id, course_id)}
    by_id = {topic.id: topic for topic in topics}

    nodes: list[KnowledgeNodeRead] = []
    for topic in topics:
        state = states.get(topic.id)
        if state is None:
            continue
        nodes.append(
            KnowledgeNodeRead(
                topic_id=topic.id,
                name=topic.name,
                description=topic.description,
                mastery=round(state.mastery_score, 1),
                effective_mastery=round(
                    effective_mastery(
                        state.mastery_score, state.evidence, state.last_practised_at
                    ),
                    1,
                ),
                band=band_for(state),
                band_label=BAND_LABELS[band_for(state)],
                questions_attempted=state.questions_attempted,
            )
        )

    rendered_edges = [
        KnowledgeEdgeRead(
            id=edge.id,
            source_topic_id=edge.source_topic_id,
            target_topic_id=edge.target_topic_id,
            relationship_type=edge.relationship_type,
            supporting_chunk_count=edge.supporting_chunk_count,
            sources=_sources_for(session, edge),
        )
        for edge in edges
        if edge.source_topic_id in by_id and edge.target_topic_id in by_id
    ]
    return nodes, rendered_edges


@router.get(
    "/courses/{course_id}/knowledge-map",
    response_model=KnowledgeMapRead,
    responses=_RESPONSES,
    summary="Read the stored knowledge map",
)
def read_knowledge_map(
    service: KnowledgeMapServiceDep,
    mastery: MasteryServiceDep,
    session: SessionDep,
    user: CurrentUser,
    course_id: uuid.UUID,
) -> KnowledgeMapRead:
    """Never generates. An empty `edges` list means the map has not been built."""
    topics, edges = service.get(user.id, course_id)
    nodes, rendered = _render_map(session, mastery, user.id, course_id, topics, edges)
    return KnowledgeMapRead(
        course_id=course_id,
        nodes=nodes,
        edges=rendered,
        generated_at=min((edge.created_at for edge in edges), default=None),
    )


@router.post(
    "/courses/{course_id}/knowledge-map",
    response_model=KnowledgeMapGenerateResponse,
    status_code=status.HTTP_201_CREATED,
    responses=_RESPONSES,
    summary="Derive the knowledge map from the course material",
)
def generate_knowledge_map(
    service: KnowledgeMapServiceDep,
    mastery: MasteryServiceDep,
    session: SessionDep,
    user: CurrentUser,
    course_id: uuid.UUID,
) -> KnowledgeMapGenerateResponse:
    result = service.generate(user.id, course_id)
    nodes, rendered = _render_map(
        session, mastery, user.id, course_id, result.topics, result.relationships
    )
    return KnowledgeMapGenerateResponse(
        course_id=course_id,
        nodes=nodes,
        edges=rendered,
        generated_at=min(
            (edge.created_at for edge in result.relationships), default=None
        ),
        candidates_rejected=result.rejected_count,
        model_calls=result.model_calls,
    )


@router.get(
    "/courses/{course_id}/knowledge-gaps",
    response_model=KnowledgeGapsRead,
    responses=_RESPONSES,
    summary="Detected knowledge gaps for this course",
)
def read_knowledge_gaps(
    service: KnowledgeMapServiceDep,
    user: CurrentUser,
    course_id: uuid.UUID,
) -> KnowledgeGapsRead:
    """Deterministic, computed from mastery and the stored graph.

    No model call happens here at all — this route works with the AI provider
    entirely unavailable, which is the point of keeping the algorithm in
    `services/learning/knowledge.py`.
    """
    _, edges = service.get(user.id, course_id)
    gaps = service.gaps(user.id, course_id)
    return KnowledgeGapsRead(
        course_id=course_id,
        gaps=[
            KnowledgeGapRead(
                topic_id=gap.topic_id,
                name=gap.name,
                kind=gap.kind,
                kind_label=GAP_KIND_LABELS[gap.kind],
                severity=gap.severity,
                effective_mastery=gap.effective_mastery,
                blocked_topics=gap.blocked_topics,
                attempted_dependents=gap.attempted_dependents,
                reason=gap.reason,
            )
            for gap in gaps
        ],
        has_map=bool(edges),
    )
